### PR TITLE
WIP: Support RKE2

### PR DESCRIPTION
**Pull request by WolfspiritM**

### CODE REVIEW

**Comment by xrow** [comment]
I would like to try this once it is merged....

**Comment by mysticaltech** [comment]
@xrow This is in the pipes! Hopefully sooner rather than later.

File: control_planes.tf [hunk 1]
```diff
@@ -71,6 +73,15 @@ resource "hcloud_load_balancer_service" "load_balancer_service" {
   listen_port      = "6443"
 }
 
+resource "hcloud_load_balancer_service" "load_balancer_service_rke2" {
```
Comment (mysticaltech) [inline control_planes.tf:76]: @WolfspiritM Do you remember what's the purpose of this?

**Comment by mysticaltech** [comment]
@captnCC Hope all is good man, long time no see. If you have bandwidth to take this on, it would be wonderful, please don't hesitate. For the rest, all is being sorted out, see the other active PR.

File: control_planes.tf (hunk 1 above)
Comment (WolfspiritM) [reply control_planes.tf:76]: RKE2 uses port 9345 to handle cluster joins. This adds the loadbalancer service for that port.

Comment (mysticaltech) [reply control_planes.tf:76]: Thanks, @WolfspiritM, if you find the courage to finish this PR one day, it would be awesome!

**Comment by mysticaltech** [comment]
@kube-hetzner/core Who wants to take this on? This repo is just for inspiration, but @WolfspiritM gave us a beautiful POC.

Both k3s and rke2 were built and are maintained by the same team over at Rancher! They are extremely similar in how they work, lot of parallels, it should be quite easy. However it does require a good amount of work and testing.

My advice, if one of you is eager to take this on, create a new branch for it, and let's go. I will help you along the way and we can make this awesome!

Supporting docs:

- https://docs.k3s.io/installation/configuration
- https://docs.rke2.io/install/configuration

**Comment by valkenburg-prevue-ch** [comment]
What are the pain points?

**Comment by ifeulner** [comment]
I can work on this, but what happened to the better kubernetes integration @valkenburg-prevue-ch ? :)

**Comment by valkenburg-prevue-ch** [comment]
I must have forgotten what I said weeks or months ago, because I'm not sure what you mean. Could you explain? 🧐

**Comment by mysticaltech** [comment]
@valkenburg-prevue-ch No paint points for me personally, just that it seems rke2 has better security, seing they are very similar, it seems like a low hanging fruit to support deploying both in this project. I will bring us closer to having something extremely solid, as rke2 is government level secure and enforces best practices.

**Comment by mysticaltech** [comment]
@ifeulner If you feel inspired please take this on!! 🙏 🚀

**Comment by valkenburg-prevue-ch** [comment]
> @valkenburg-prevue-ch No paint points for me personally, ...

I meant more: what is needed, what needs to be fixed/checked/tested? I mean, I'm eager to adopt RKE2, but my current priorities are elsewhere.

**Comment by zifeo** [comment]
I also would love to see rke2 support for Hertzner but do not have the bandwidth currently. Happy to help review though or to bring some inspirations over from the OpenStack [version](https://github.com/zifeo/terraform-openstack-rke2) I created aside. Let me know!

**Comment by valkenburg-prevue-ch** [comment]
I've tangentially been looking into this (not putting in serious work), and I think that the selinux problem which we had with k3s which was later resolved by an update on k3s-selinux, has not yet been fixed for rke2-selinux. See https://github.com/rancher/rke2-selinux (no updates since november) and https://bugzilla.opensuse.org/show_bug.cgi?id=1207836 (issue is the same for k3s and rke2).

So, at this moment, rke2 on microos would only work without full selinux setup.

**Comment by valkenburg-prevue-ch** [comment]
And, by the way, the minimum hardware for a rke2 node is considerably larger than k3s:

## k3s 
Recommended 2 CPUs and 1GB RAM (https://docs.k3s.io/installation/requirements#hardware), that's a CPX11 Hetzner server.

## rke2
Recommended 4CPUs and 8GB RAM (https://docs.rke2.io/install/requirements#hardware), that's a CPX31 Hetzner server, a bit over 3x the price of a CPX11.

**Comment by mysticaltech** [comment]
@valkenburg-prevue-ch Very good points! I had no idea they were behind on the SELinux issue, that is a bad sign, it means it's used a lot less.

About the node requirements, I was aware it was more, but not that much more.

In that new light, I could live without rke2, @ifeulner, the ball is in your camp. If you still want to make it work, we could probably fix the `rke2_selinux` package ourselves based on the changes done on the k3s one, I already compiled the latter RPM once and it's not sorcery.

**Comment by ifeulner** [comment]
@valkenburg-prevue-ch good catch with the selinux topic... and a pity that it not has been fixed yet.
Regarding node requirements: Yes, was aware of that, but to be honest, none of our cluster has nodes with less than 16GiB RAM (at least for the agents), so this is not a topic for me (and most who want to have a production-grade cluster, I think). 
RKE2 has a close alignment to upstream kubernetes and is more secure including FIPS140-2 if setup correspondingly. that supports a lot of governmental use cases, that's why I think it's attractive to support it.

**Comment by valkenburg-prevue-ch** [comment]
Yes, I agree, @ifeulner . 

I've been wondering if, since the motivation is security, it would make sense to go for an architecture where none of the cluster nodes have a public ip, and all maintenance and node-level egress pass through a bastion. I think `squid` is a good candidate for running a proxy in the bastion, listening only on private network.

**Comment by mysticaltech** [comment]
@valkenburg-prevue-ch Basically in the current setup, [you can close the SSH port](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/436#issuecomment-1379241029) after the cluster is up and that basically isolates it, and same for the other in ports like the Kubernetes API one.

**Comment by mysticaltech** [comment]
FYI I went ahead and let the admins of `rke2_selinux` know of our intention to use it and our need for the fix. https://github.com/rancher/rke2-selinux/issues/33

**Comment by ifeulner** [comment]
> I've been wondering if, since the motivation is security, it would make sense to go for an architecture where none of the cluster nodes have a public ip, and all maintenance and node-level egress pass through a bastion. I think `squid` is a good candidate for running a proxy in the bastion, listening only on private network.

I am not sure how much more security that gives you compared to block the port via a firewall. 
But it introduces a lot of new challenges... without a public interface the host has no internet connectivity. So you have either to use a proxy for every outside connection or a router. These resources need to be managed / maintained and introduce new bottlenecks / overhead.

**Comment by valkenburg-prevue-ch** [comment]
That's true. GKE has a NAT router for their solution, https://cloud.google.com/kubernetes-engine/docs/concepts/private-cluster-concept . The fact that it exists means there are customers who appreciate this design. It's how you do it when you setup a classical hadoop cluster for banks. But yes, as Hetzner does not have NAT routers in their cloud offering (or am I wrong?) you have to have a proxy on the bastion (or another) host, like I said already. 

I understand this might be too heavy on maintenance on this public open source project.

**Comment by valkenburg-prevue-ch** [comment]
About the security:
- it's simply a different attack surface compared the firewall-only solution,
- with a proxy like squid, you can monitor (and filter) all outgoing traffic.

**Comment by jspizziri** [comment]
We're currently running rke2 on AWS and have been playing around with the idea of migrating to Hetzner. I can't guarantee anything but we might be able to justify pushing this PR forward to explore that further. I'd be interested in what the maintainers Definition of Done is for merge. I haven't given the TF a thorough look yet but as I understand it from the above:

- [ ] Fix the `rke2-selinux` bug
- [ ] ...?

Is it really just a matter of taking the PR and trying to get something deployed on Hetzner with it? Thanks in advance!

**Comment by mysticaltech** [comment]
@jspizziri Thanks for looking into this. Basically, this PR will be closed as no longer up to date with the changes. However, we will try to support RKE2. And that should be straightforward because we use k3s, and RKE2 is done by the same team using the the same config format and flow.

Done will be reached when a future PR (that you could start if you want) has both k3s and RKE2 well modularized and working with most features that this project offers.

The rke2-selinux is not something I worry a lot, because it can be fixed in a few hours of testing, it's just selinux config files, and a similar fix has already been done on the rancher k3s-selinux repo.

**Pull request closed without merging**